In [1]:
# IMPORTING OF LIBRARIES
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# IMPORT DATASETS

In [2]:
df = pd.read_csv('HOUSING_PRICE.csv')

In [3]:
df.shape

(1460, 81)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

# SEPARATE INT,FLOAT DATA TYPE COLUMNS FROM OBJECT COLUMNS

In [5]:
# Select object data type columns
df1 = df.select_dtypes(include=['object'])

# Select non-object data type columns
df = df.select_dtypes(exclude=['object'])

print("df1:")
print(df1)
print("\ndf:")
print(df)

df1:
     MSZoning Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
0          RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1          RL   Pave   NaN      Reg         Lvl    AllPub       FR2       Gtl   
2          RL   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   
3          RL   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl   
4          RL   Pave   NaN      IR1         Lvl    AllPub       FR2       Gtl   
...       ...    ...   ...      ...         ...       ...       ...       ...   
1455       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1456       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1457       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1458       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1459       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

     Neighborhood Cond

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 38 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   LotFrontage    1201 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   OverallQual    1460 non-null   int64  
 5   OverallCond    1460 non-null   int64  
 6   YearBuilt      1460 non-null   int64  
 7   YearRemodAdd   1460 non-null   int64  
 8   MasVnrArea     1452 non-null   float64
 9   BsmtFinSF1     1460 non-null   int64  
 10  BsmtFinSF2     1460 non-null   int64  
 11  BsmtUnfSF      1460 non-null   int64  
 12  TotalBsmtSF    1460 non-null   int64  
 13  1stFlrSF       1460 non-null   int64  
 14  2ndFlrSF       1460 non-null   int64  
 15  LowQualFinSF   1460 non-null   int64  
 16  GrLivArea      1460 non-null   int64  
 17  BsmtFullBath   1460 non-null   int64  
 18  BsmtHalf

# CHECK & REMOVE DUPLICATES

In [6]:
df.duplicated().sum()

0

# CHECK & IMPUTE NULL VALUES

In [7]:
df.isnull().sum()

Id                 0
MSSubClass         0
LotFrontage      259
LotArea            0
OverallQual        0
OverallCond        0
YearBuilt          0
YearRemodAdd       0
MasVnrArea         8
BsmtFinSF1         0
BsmtFinSF2         0
BsmtUnfSF          0
TotalBsmtSF        0
1stFlrSF           0
2ndFlrSF           0
LowQualFinSF       0
GrLivArea          0
BsmtFullBath       0
BsmtHalfBath       0
FullBath           0
HalfBath           0
BedroomAbvGr       0
KitchenAbvGr       0
TotRmsAbvGrd       0
Fireplaces         0
GarageYrBlt       81
GarageCars         0
GarageArea         0
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
MiscVal            0
MoSold             0
YrSold             0
SalePrice          0
dtype: int64

In [8]:
df['LotFrontage'].fillna(df['LotFrontage'].mean(), inplace=True)
df['MasVnrArea'].fillna(df['MasVnrArea'].mean(), inplace=True)
df['GarageYrBlt'].fillna(df['GarageYrBlt'].mean(), inplace=True)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 38 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   LotFrontage    1460 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   OverallQual    1460 non-null   int64  
 5   OverallCond    1460 non-null   int64  
 6   YearBuilt      1460 non-null   int64  
 7   YearRemodAdd   1460 non-null   int64  
 8   MasVnrArea     1460 non-null   float64
 9   BsmtFinSF1     1460 non-null   int64  
 10  BsmtFinSF2     1460 non-null   int64  
 11  BsmtUnfSF      1460 non-null   int64  
 12  TotalBsmtSF    1460 non-null   int64  
 13  1stFlrSF       1460 non-null   int64  
 14  2ndFlrSF       1460 non-null   int64  
 15  LowQualFinSF   1460 non-null   int64  
 16  GrLivArea      1460 non-null   int64  
 17  BsmtFullBath   1460 non-null   int64  
 18  BsmtHalf

In [10]:
df.describe().T.round(0)

count      mean      std      min       25%       50%  \
Id             1460.0     730.0    422.0      1.0     366.0     730.0   
MSSubClass     1460.0      57.0     42.0     20.0      20.0      50.0   
LotFrontage    1460.0      70.0     22.0     21.0      60.0      70.0   
LotArea        1460.0   10517.0   9981.0   1300.0    7554.0    9478.0   
OverallQual    1460.0       6.0      1.0      1.0       5.0       6.0   
OverallCond    1460.0       6.0      1.0      1.0       5.0       5.0   
YearBuilt      1460.0    1971.0     30.0   1872.0    1954.0    1973.0   
YearRemodAdd   1460.0    1985.0     21.0   1950.0    1967.0    1994.0   
MasVnrArea     1460.0     104.0    181.0      0.0       0.0       0.0   
BsmtFinSF1     1460.0     444.0    456.0      0.0       0.0     384.0   
BsmtFinSF2     1460.0      47.0    161.0      0.0       0.0       0.0   
BsmtUnfSF      1460.0     567.0    442.0      0.0     223.0     478.0   
TotalBsmtSF    1460.0    1057.0    439.0      0.0     796.0     992.0   
1stFlrSF       1460.0    1163.0    387.0    334.0     882.0    1087.0   
2ndFlrSF       1460.0     347.0    437.0      0.0       0.0       0.0   
LowQualFinSF   1460.0       6.0     49.0      0.0       0.0       0.0   
GrLivArea      1460.0    1515.0    525.0    334.0    1130.0    1464.0   
BsmtFullBath   1460.0       0.0      1.0      0.0       0.0       0.0   
BsmtHalfBath   1460.0       0.0      0.0      0.0       0.0       0.0   
FullBath       1460.0       2.0      1.0      0.0       1.0       2.0   
HalfBath       1460.0       0.0      1.0      0.0       0.0       0.0   
BedroomAbvGr   1460.0       3.0      1.0      0.0       2.0       3.0   
KitchenAbvGr   1460.0       1.0      0.0      0.0       1.0       1.0   
TotRmsAbvGrd   1460.0       7.0      2.0      2.0       5.0       6.0   
Fireplaces     1460.0       1.0      1.0      0.0       0.0       1.0   
GarageYrBlt    1460.0    1979.0     24.0   1900.0    1962.0    1979.0   
GarageCars     1460.0       2.0      1.0      0.0       1.0       2.0   
GarageArea     1460.0     473.0    214.0      0.0     334.0     480.0   
WoodDeckSF     1460.0      94.0    125.0      0.0       0.0       0.0   
OpenPorchSF    1460.0      47.0     66.0      0.0       0.0      25.0   
EnclosedPorch  1460.0      22.0     61.0      0.0       0.0       0.0   
3SsnPorch      1460.0       3.0     29.0      0.0       0.0       0.0   
ScreenPorch    1460.0      15.0     56.0      0.0       0.0       0.0   
PoolArea       1460.0       3.0     40.0      0.0       0.0       0.0   
MiscVal        1460.0      43.0    496.0      0.0       0.0       0.0   
MoSold         1460.0       6.0      3.0      1.0       5.0       6.0   
YrSold         1460.0    2008.0      1.0   2006.0    2007.0    2008.0   
SalePrice      1460.0  180921.0  79443.0  34900.0  129975.0  163000.0   

                    75%       max  
Id               1095.0    1460.0  
MSSubClass         70.0     190.0  
LotFrontage        79.0     313.0  
LotArea         11602.0  215245.0  
OverallQual         7.0      10.0  
OverallCond         6.0       9.0  
YearBuilt        2000.0    2010.0  
YearRemodAdd     2004.0    2010.0  
MasVnrArea        164.0    1600.0  
BsmtFinSF1        712.0    5644.0  
BsmtFinSF2          0.0    1474.0  
BsmtUnfSF         808.0    2336.0  
TotalBsmtSF      1298.0    6110.0  
1stFlrSF         1391.0    4692.0  
2ndFlrSF          728.0    2065.0  
LowQualFinSF        0.0     572.0  
GrLivArea        1777.0    5642.0  
BsmtFullBath        1.0       3.0  
BsmtHalfBath        0.0       2.0  
FullBath            2.0       3.0  
HalfBath            1.0       2.0  
BedroomAbvGr        3.0       8.0  
KitchenAbvGr        1.0       3.0  
TotRmsAbvGrd        7.0      14.0  
Fireplaces          1.0       3.0  
GarageYrBlt      2001.0    2010.0  
GarageCars          2.0       4.0  
GarageArea        576.0    1418.0  
WoodDeckSF        168.0     857.0  
OpenPorchSF        68.0     547.0  
EnclosedPorch       0.0     552.0  
3SsnPorch       

# DEFINING FEATURES & TARGETS

In [11]:
X = df.drop(columns=['SalePrice'])
Y = df['SalePrice']
X.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  GarageArea  WoodDeckSF  \
0          2003       196.0         706  ...         548           0   
1          1976         0.0         978  ...         460         298   
2          2002       162.0         486  ...         608           0   
3          1970         0.0         216  ...         642           0   
4          2000       350.0         655  ...         836         192   

   OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  \
0           61              0          0            0         0        0   
1            0              0          0            0         0        0   
2           42              0          0            0         0        0   
3           35            272          0            0         0        0   
4           84              0          0            0         0        0   

   MoSold  YrSold  
0       2    2008  
1       5    2007  
2       9    2008  
3       2    2006  
4      12    2008  

[5 rows x 37 columns]

In [12]:
Y

0       208500
1       181500
2       223500
3       140000
4       250000
         ...  
1455    175000
1456    210000
1457    266500
1458    142125
1459    147500
Name: SalePrice, Length: 1460, dtype: int64

In [13]:
X.shape,Y.shape

((1460, 37), (1460,))

# SAMPLING

In [14]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=0)

In [15]:
X_train.shape,X_test.shape,Y_train.shape,Y_test.shape

((1168, 37), (292, 37), (1168,), (292,))

In [16]:
X_train.shape[0] / X.shape[0],X_test.shape[0] / X.shape[0],Y_train.shape[0]/Y.shape[0],Y_test.shape[0]/Y.shape[0]

(0.8, 0.2, 0.8, 0.2)

# MODEL FITTING
### LINEAR REGRESSION MODEL

In [17]:
from sklearn.linear_model import LinearRegression

In [18]:
mymodel = LinearRegression()

In [19]:
mymodel.fit(X_train,Y_train)

LinearRegression()

# IMPLIMENTATION OF METRICS MSE, RMSE, NRMSE , R2

In [20]:
from sklearn.metrics import mean_squared_error, r2_score

# METRICS VALUE FOR TRAIN TARGET

In [21]:
Y_train_predicted = mymodel.predict(X_train)

In [22]:
MSE_Train = mean_squared_error(Y_train,Y_train_predicted).round()
MSE_Train

938163782.0

In [24]:
RMSE = (MSE_Train**0.5).round()
RMSE

30629.0

In [45]:
NRMSE = RMSE /720100
NRMSE.round(3)

0.043

In [46]:
r2_score(Y_train,Y_train_predicted).round(3)

0.848

CONOCLISION:- 1. NRMSE is low 0.043 (o.2 to 0.5) , which implies that the model is a stable model. 2. R2 is high 
0.848 (~ between 0.5 to 0.99) which implies the model is very stable model.

# METRICS VALUE FOR TEST TARGETS

In [47]:
Y_test_predicted = mymodel.predict(X_test)

In [48]:
MSE_Test = mean_squared_error(Y_test,Y_test_predicted).round()
MSE_Test

2527383335.0

In [51]:
RMSE = (MSE_Test**0.5).round()
RMSE

50273.0

In [52]:
NRMSE = RMSE /720100
NRMSE.round(3)

0.07

In [53]:
r2_score(Y_test,Y_test_predicted).round(3)

0.634

CONCLUSION:- 1. NRMSE is low 0.07 (o.2 to 0.5) , which implies that the model is a stable model. 2. R2 is high 
0.634 (~ between 0.5 to 0.99) which implies the model is very stable model.

***********************************DONE************************************************